Documentation/llilc-gc-transition.md

```diff
@@ -0,0 +1,96 @@
+# GC Transition Support in LLILC
+
+The CoreCLR runtime supports calls from managed (i.e. GC-aware) code to
+unmanaged (i.e. GC-unaware) code. In the runtime's parlance, such calls are
+termed "P/Invokes" (short for Platform Invoke); in LLILC/LLVM they are referred
+to as "unmanaged calls" or "GC transitions". Because the target of such a call
+is not aware of the CoreCLR GC, such calls require runtime-specific setup and
+teardown code. The nature of this transition code requires code in both LLILC
+and LLVM. A brief discussion of the specifics of GC transitions in CoreCLR and
+the support for these transitions in LLILC and LLVM follows.
+
+## GC transitions in CoreCLR
+
+In order for the CoreCLR garbage collector to perform garbage collection it must
+be able to
+1. logically pause all threads that are running managed code, and
+2. scan each thread's stack for GC references.
+
+### Thread suspension and GC modes
+
+To meet the first requirement above without physically suspending arbitrary
+threads (and thus aggravating the chance of deadlocks in the runtime), each
+thread that the runtime is tracking is assigned one of two modes: cooperative
+mode and preemptive mode.
+
+#### Cooperative mode
+
+Threads in cooperative mode are considered to be running managed code. These
+threads are therefore potentially modifying GC referencs, and must cooperate
+with the runtime in order to reach a GC safepoint before a collection may
+occur. All managed code (and almost JIT compiled code will run in cooperative
+mode.
+
+#### Preemptive mode
+
+Threads in preemprive mode are considered to be running unmanaged code. Such a
+thread obeys three invariants:
+1. It must not currently be manipulating GC references.
```

Comment:
To make (1) clearer, it may be better to split into two:
(1a) Must not modify gc-pointers
(1b) All Gc-pointer roots must be reachable from managed stack-frames

(1b) makes the requirement for spilling gc-pointers in callee-save registers explicit.